### PR TITLE
Allow an investment project summary to be edited

### PR DESCRIPTION
**Pull request by teneightfive**
Adds the ability to edit the project summary screen and update it.

### CODE REVIEW

File: src/controllers/investment-create.controller.js [hunk 1]
```diff
@@ -1,24 +1,29 @@
...
+  getInvestmentProjectSummary,
+  saveInvestmentProject,
+  editInvestmentProject
+} = require('../repos/investment.repo')
+const { transformForApi, transformFromApi } = require('../services/investment-formatting.service')
 const metadataRepo = require('../repos/metadata.repo')
 
 function getHandler (req, res, next) {
-  const equityCompanyId = req.query['equity-company']
+  const equityCompanyId = res.locals.equityCompany || req.query['equity-company']
```
Comment (feedmypixel) [inline src/controllers/investment-create.controller.js]: do we need to support both?

File: src/controllers/investment-create.controller.js [hunk 2]
```diff
@@ -63,14 +75,32 @@ function postHandler (req, res, next) {
...
+}
+
 router.get('/investment/create', getHandler)
 router.post('/investment/create', postHandler, getHandler)
 
-module.exports = { router, getHandler, postHandler }
+router.get('/investment/:id/details/edit', editHandler, getHandler)
+router.post('/investment/:id/details/edit', editHandler, postHandler, getHandler)
+
+module.exports = { router, getHandler, postHandler, editHandler }
```
Comment (feedmypixel) [inline src/controllers/investment-create.controller.js]: newlines for these?

File: src/services/investment-formatting.service.js [hunk 3]
```diff
@@ -78,6 +84,8 @@ function transformForApi (body) {
     project['referral_source_advisor'] = body['is-referral-source']
   }
 
+  project['estimated_land_date'] = `${body['land-date_year']}-${body['land-date_month']}-01`
+
   Object.keys(project).forEach((key) => {
```
Comment (feedmypixel) [inline src/services/investment-formatting.service.js]: Not part of this PR but it feels like the data needs to be normalised to either an array or an object, not both.

File: src/services/investment-formatting.service.js [hunk 4]
```diff
@@ -87,22 +95,56 @@ function transformForApi (body) {
       project[key] = [{
         id: project[key]
       }]
+    } else if (remove.includes(key)) {
```
Comment (feedmypixel) [inline src/services/investment-formatting.service.js]: can you remove unwanted keys before they get iterated over?

File: src/services/investment-formatting.service.js [hunk 5]
```diff
@@ -87,22 +95,56 @@ function transformForApi (body) {
...
 
-  delete project['land-date_year']
-  delete project['land-date_month']
-  delete project['is-relationship-manager']
-  delete project['is-referral-source']
+  const date = new Date(formatted['estimated_land_date'])
+  formatted['land-date_year'] = date.getFullYear()
+  formatted['land-date_month'] = date.getMonth() + 1 // month is zero based index
+
+  Object.keys(formatted).forEach((key) => {
```
Comment (feedmypixel) [inline src/services/investment-formatting.service.js]: This is also being used above - can this be abstracted into a function?

File: src/services/investment-formatting.service.js [hunk 6]
```diff
@@ -87,22 +95,56 @@ function transformForApi (body) {
...
+  return formatted
 }
 
 module.exports = {
   getInvestmentDetailsDisplay,
   getOpenInvestmentProjects,
   getClosedInvestmentProjects,
-  transformForApi
+  transformForApi,
+  transformFromApi
```
Comment (feedmypixel) [inline src/services/investment-formatting.service.js]: the wording of these methods are very similar and I'm not sure what they do or how they are different. Could a rename make that clearer?

Comment (feedmypixel) [inline src/services/investment-formatting.service.js]: If I have understood right maybe these are clearer and less likely to be mixed up?
- ` transformToApi`
- `transformFromApi`

File: src/views/investment/create.njk [hunk 7]
```diff
@@ -2,27 +2,32 @@
...
+    <table class="table--key-value">
+      <tbody>
+        <tr>
+          <td>{{ equityCompany.name }}</td>
+        </tr>
+        <tr>
+          <td>{{ equityCompany.registered_address_country.name }}</td>
+        </tr>
+        <tr>
+          <td>{{ equityCompanyInvestments.count|default('No', true) }} investment project{{ 's' if equityCompanyInvestments.count != 1 }} in the UK</td>
```
Comment (feedmypixel) [inline src/views/investment/create.njk]: FYI there is [pluralise filter](https://github.com/uktrade/data-hub-fe-beta2/pull/139/files#diff-a34f9a72bb41fd9597d5ec8aa14d4e1fR18) coming

Comment (teneightfive) [inline src/views/investment/create.njk]: _applaudes_

File: src/controllers/investment-create.controller.js (hunk 1 above)
Comment (teneightfive) [inline src/controllers/investment-create.controller.js]: Reason being is one is coming from the `editHandler` middleware (which thinking about it, I may rename to `editMiddleware` to be more clear) when editing a project summary, and the other is being set as a querystring param. 

When editing I don't really want to redirect to the get with the querystring as you'd lose the values in `res.locals`.

I could refactor so that there is a middleware on all routes that looks for the querystring params and also sets it on to `res.locals` so that this function only expects that value to come from one place?

File: src/services/investment-formatting.service.js (hunk 3 above)
Comment (teneightfive) [inline src/services/investment-formatting.service.js]: I don't really like what these functions are doing to be honest. It's compensating for the way the API sends that data back. We have to expand a string into an object in some cases where I think it might be nicer for the API to accept a string or an object, and if a string, treat it as the ID value.

File: src/services/investment-formatting.service.js (hunk 4 above)
Comment (teneightfive) [inline src/services/investment-formatting.service.js]: Yes that's probably more efficient.

File: src/repos/investment.repo.js [hunk 8]
```diff
@@ -17,11 +17,19 @@ function getInvestmentRequirements (token, investmentId) {
   return authorisedRequest(token, `${config.apiRoot}/v3/investment/${investmentId}/requirements`)
 }
 
-function saveInvestmentProject (token, project) {
+function saveInvestmentProject (token, body) {
```
Comment (tyom) [inline src/repos/investment.repo.js]: I wonder if create/update would be more appropriate verbs.

File: src/controllers/investment-create.controller.js (hunk 1 above)
Comment (feedmypixel) [inline src/controllers/investment-create.controller.js]: >When editing I don't really want to redirect to the get with the querystring as you'd lose the values in res.locals.

sounds sensible

>I could refactor so that there is a middleware on all routes that looks for the querystring params and also sets it on to res.locals so that this function only expects that value to come from one place?

Feels like you would expect a query to be on req.query rather than res.locals. Which would make it a little magic IMO. 

It sounds like for the moment we have two ways in.

Maybe this is a bigger thing in that we have a `getHandler` which is doing the get work for two things. Render a company page and edit a company page. Potentially we have a handler for the display, and then if it is an edit it falls through to an edit handler?

*caveat I have not had my head in the code so take whichever route you feel is sensible*

File: src/views/investment/create.njk (hunk 7 above)
Comment (feedmypixel) [inline src/views/investment/create.njk]: 🙇

File: src/controllers/investment-create.controller.js (hunk 1 above)
Comment (teneightfive) [inline src/controllers/investment-create.controller.js]: The reason for using the same handler is there is next to no difference apart from where the equity company ID is coming from, hence the thought to use the same handler but have the edit route add an additional piece of middleware.

Let me see if I can refactor it a bit.

Comment (feedmypixel) [inline src/controllers/investment-create.controller.js]: I quite like the middleware pattern on routes. If it doesn't do something in the first handler just drop it through to the next, same principle as the error handling.

This approach also works really nicely when doing errors with `post -> redirect get`

Comment (teneightfive) [inline src/controllers/investment-create.controller.js]: Yep, am doing that with the post/get actions. The post route includes the postaction and then drops through to the get action. Does just mean you have to set things on `res.locals` though.

Comment (teneightfive) [inline src/controllers/investment-create.controller.js]: Had a chat with @tyom and for now I think we're going to leave it rather than adding more middleware. It's something we'll probably revisit when we look to combine a lot of the add/edit flows that we currently have.
